lib/Sema/TypeCheckGeneric.cpp

```diff
@@ -616,32 +616,124 @@ static void checkReferencedGenericParams(GenericContext *dc,
 
   auto *decl = cast<ValueDecl>(dc->getInnermostDeclarationDeclContext());
 
-  // Collect all generic params referenced in parameter types,
-  // return type or requirements.
-  SmallPtrSet<GenericTypeParamDecl *, 4> referencedGenericParams;
+  // A helper class to collect referenced generic type parameters
+  // and dependent memebr types.
+  class ReferencedGenericTypeWalker : public TypeWalker {
+    SmallPtrSet<CanType, 4> ReferencedGenericParams;
+
+  public:
+    ReferencedGenericTypeWalker() {}
+    Action walkToTypePre(Type ty) override {
+      // Find generic parameters or dependent member types.
+      // Once such a type is found, don't recurse into its children.
+      if (ty->hasTypeParameter()) {
+        if (ty->isTypeParameter()) {
+          ReferencedGenericParams.insert(ty->getCanonicalType());
+          return Action::SkipChildren;
+        }
+        return Action::Continue;
+      }
+      return Action::SkipChildren;
+    }
 
-  auto visitorFn = [&referencedGenericParams](Type t) {
-    if (auto *paramTy = t->getAs<GenericTypeParamType>())
-      referencedGenericParams.insert(paramTy->getDecl());
+    SmallPtrSet<CanType, 4> &getReferencedGenericParams() {
+      return ReferencedGenericParams;
+    }
   };
 
+  // Collect all generic params referenced in parameter types and
+  // return type.
+  ReferencedGenericTypeWalker ParamsAndResultWalker;
   auto *funcTy = decl->getInterfaceType()->castTo<GenericFunctionType>();
-  funcTy->getInput().visit(visitorFn);
-  funcTy->getResult().visit(visitorFn);
+  funcTy->getInput().walk(ParamsAndResultWalker);
+  funcTy->getResult().walk(ParamsAndResultWalker);
+
+  // Set of generic params referenced in parameter types,
+  // return type or requirements.
+  auto &referencedGenericParams =
+      ParamsAndResultWalker.getReferencedGenericParams();
+
+  // Check if at least one of the generic params in the requirment refers
+  // to an already referenced generic parameter. If this is the case,
+  // then the other type is also considered as referenced, because
+  // it is used to put requirements on the first type.
+  auto reqTypesVisitor = [&referencedGenericParams](Requirement req) -> bool {
+    Type first;
+    Type second;
 
-  auto requirements = sig->getRequirements();
-  for (auto req : requirements) {
     switch (req.getKind()) {
-      case RequirementKind::SameType:
-      case RequirementKind::Superclass:
-        req.getSecondType().visit(visitorFn);
-        LLVM_FALLTHROUGH;
-
-      case RequirementKind::Conformance:
-      case RequirementKind::Layout:
-        req.getFirstType().visit(visitorFn);
-        break;
+    case RequirementKind::SameType:
+      second = req.getSecondType();
+      LLVM_FALLTHROUGH;
+
+    case RequirementKind::Superclass:
```

Comment:
Shouldn't we also capture the second type of superclass constraints, if only for future-correctness?